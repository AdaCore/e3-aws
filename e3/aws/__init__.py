import argparse
from uuid import uuid4
from botocore.stub import Stubber
from e3.env import Env
import botocore.session


class Session(object):
    """Handle AWS session and clients."""

    def __init__(self, regions=None, stub=False, profile=None,
                 credentials=None):
        """Initialize an AWS session.

        Once intialized AWS environment can be accessed from Env().aws_env

        :param regions: list of regions to work on. The first region is
            considered as the default region.
        :type regions: list[str]
        :param stub: if True clients are necessarily stubbed
        :type stub: bool
        :param profile: profile name
        :type profile: str | None
        :param credentials: AWS credentials dictionary containing the
            following keys: AccessKeyId, SecretAccessKey, SessionToken
            as returnedy by ``assume_role``
        :type credentials: dict[str]
        """
        if profile is not None or credentials is None:
            self.session = botocore.session.Session(profile=profile)
        else:
            self.session = botocore.session.Session()
            self.session.set_credentials(
                access_key=credentials['AccessKeyId'],
                secret_key=credentials['SecretAccessKey'],
                token=credentials['SessionToken'])

        self.profile = profile
        if regions is None:
            self.regions = [self.session.region_name]
        else:
            self.regions = regions
        self.default_region = None
        self.force_stub = stub
        self.clients = {}
        self.stubbers = {}

        self._account_alias = None

    def assume_role(self, role_arn, role_session_name):
        """Return a session with ``role_arn`` credentials.

        :param role_arn: ARN of the role to assume
        :type role_arn: str
        :param role_session_name: a name to associate with the created
            session
        :type role_session_name: str

        :return: a Session instance
        :rtype: Session
        """
        credentials = self.assume_role_get_credentials(
            role_arn, role_session_name)
        return Session(regions=self.regions, credentials=credentials)

    def assume_role_get_credentials(self, role_arn, role_session_name):
        """Return credentials for ``role_arn``.

        :param role_arn: ARN of the role to assume
        :type role_arn: str
        :param role_session_name: a name to associate with the created
            session
        :type role_session_name: str

        :return: credentials dictionary
        :rtype: dict
        """
        client = self.client('sts', region=self.regions[0])
        response = client.assume_role(
            RoleArn=role_arn,
            RoleSessionName=role_session_name)
        return response['Credentials']

    @property
    def account_alias(self):
        """Return current account alias."""
        if self._account_alias is None:
            client = self.client('iam', region='us-east-1')
            aliases = client.list_account_aliases()['AccountAliases']
            if aliases:
                # Even if the API return a list there currently only one
                # alias possible
                self._account_alias = aliases[0]
            else:
                self._account_alias = ''
        return self._account_alias

    def stub(self, name, region=None):
        """Return stub for a given client.

        Note that if the client does not exist yet it will be created.

        :param name: client name
        :type name: str
        :param region: region associated with the client. If None the default
            region is taken.
        :type region: str | None
        :return: the stub instance
        :rtype: botocore.stub.Stubber
        """
        if not self.force_stub:
            return None
        if region is None:
            region = self.default_region

        if name not in self.stubbers or region not in self.stubbers[name]:
            # Create client
            self.client(name, region)

        return self.stubbers[name][region]

    def client(self, name, region=None):
        """Get a client.

        :param name: client name
        :type name: str
        :param region: region associated with the client. If None the default
            region is taken.
        :type region: str | None
        :return: a client instance
        :rtype: botocore.Client
        """
        if region is None:
            region = self.default_region

        assert region is not None, 'no region or default_region set'

        if name not in self.clients:
            self.clients[name] = {}
            self.stubbers[name] = {}

        if region not in self.clients[name]:
            self.clients[name][region] = \
                self.session.create_client(name, region_name=region)
            if self.force_stub:
                self.stubbers[name][region] = \
                    Stubber(self.clients[name][region])
                self.stubbers[name][region].activate()

        return self.clients[name][region]


class AWSEnv(Session):
    """Handle AWS session and clients."""

    def __init__(self, regions=None, stub=False, profile=None):
        """Initialize an AWS session.

        Once intialized AWS environment can be accessed from Env().aws_env

        :param regions: list of regions to work on. The first region is
            considered as the default region.
        :type regions: list[str]
        :param stub: if True clients are necessarily stubbed
        :type stub: bool
        :param profile: profile name
        :type profile: str | None
        """
        super().__init__(regions=regions, stub=stub, profile=profile)
        env = Env()
        env.aws_env = self


class default_region(object):
    """Context manager used to set a default region."""

    def __init__(self, region):
        """Initialize context manager.

        :param region: default region
        :type region: str
        """
        aws_env = Env().aws_env

        self.previous_region = aws_env.default_region
        self.default_region = region

    def __enter__(self):
        Env().aws_env.default_region = self.default_region

    def __exit__(self, _type, _value, _tb):
        del _type, _value, _tb
        Env().aws_env.default_region = self.previous_region


def client(name):
    """Decorate a function to handle automatically AWS client retrieval.

    The function in input should take a mandatory argument called client.
    The function seen by the user will have an optional argument region
    to select the region in which the client is created.

    :param name: client name
    :type name: str
    """
    def decorator(func):
        def wrapper(*args, **kwargs):
            aws_env = Env().aws_env
            if 'region' in kwargs:
                region = kwargs['region']
                del kwargs['region']
            else:
                region = aws_env.default_region
            client = aws_env.client(name, region=region)
            return func(*args, client=client, **kwargs)
        return wrapper
    return decorator


def session():
    """Decorate a function to handle automatically AWS session retrieval.

    The function in input should take a mandatory argument called session.

    :param name: session name
    :type name: str
    """
    def decorator(func):
        def wrapper(*args, **kwargs):
            if 'session' in kwargs:
                session = kwargs.get('session')
                del kwargs['session']
            else:
                session = Env().aws_env
            return func(*args, session=session, **kwargs)
        return wrapper
    return decorator


def assume_role_main():
    """Generate shell commands to set credentials for a role."""
    argument_parser = argparse.ArgumentParser()
    argument_parser.add_argument("role_arn")
    args = argument_parser.parse_args()

    key_to_envvar = {
        'AccessKeyId': 'AWS_ACCESS_KEY_ID',
        'SecretAccessKey': 'AWS_SECRET_ACCESS_KEY',
        'SessionToken': 'AWS_SESSION_TOKEN'}

    s = Session(regions=['eu-west-1'])
    for k, v in s.assume_role_get_credentials(
        args.role_arn,
            str(uuid4()).replace('-', '')).items():
        if k in key_to_envvar:
            print('export {}={}'.format(key_to_envvar[k], v))
