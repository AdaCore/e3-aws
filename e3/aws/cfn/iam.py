import abc

from e3.aws.cfn import AWSType, GetAtt, Resource, Stack


class Statement(object, metaclass=abc.ABCMeta):
    """Statement of IAM Policy Document."""

    def __init__(self,
                 sid=None,
                 to=None,
                 on=None,
                 not_on=None,
                 service=None):
        """Initialize a statement.

        :param sid: statement id (optional)
        :type sid: str
        :param to: one or several action for the statement
        :type to: list[str] | str | None
        :param on: resource or list of resources on which the statement apply
        :type on: list[str] | str | None
        :param not_on: resource or list of resources on which the statement
            does not apply. Note that not_on and on cannot be both set
        :type not_on: list[str] | str | None
        :param services: service of list of services that can use that
            statement
        :type services: list[str] | str | None
        """
        self.resources = []
        self.not_resources = []
        self.actions = []
        self.principals = {}
        self.sid = sid

        if to is not None:
            self.to(to)
        if on is not None:
            self.on(on)
        if not_on is not None:
            self.not_on(not_on)
        if service is not None:
            self.service(service)

    @property
    def properties(self):
        """Serialize the object as a simple dict.

        Can be used to transform to CloudFormation Yaml format.

        :rtype: dict
        """
        result = {'Effect': self.EFFECT}
        if self.sid is not None:
            result['Sid'] = self.sid
        if self.resources:
            result['Resource'] = self.resources
        if self.not_resources:
            result['NotResource'] = self.not_resources
        if self.principals:
            result['Principal'] = self.principals
        if self.actions:
            result['Action'] = self.actions
        return result

    def to(self, actions):
        """Add action(s) targeted by statement.

        :param actions: one or several action for the statement
        :type actions: collections.Iterable[str] | str | GetAtt
        :return: the modified statement
        :rtype: Statement
        """
        if isinstance(actions, str) or isinstance(actions, GetAtt):
            self.actions.append(actions)
        else:
            self.actions += actions
        return self

    def on(self, resources):
        """Add resource(s) on which the statement apply.

        :param resources: resource or list of resources
        :type resources: list[str] | str | GetAtt
        :return: the modified statement
        :rtype: Statement
        """
        assert not self.not_resources
        if isinstance(resources, str) or isinstance(resources, GetAtt):
            resources = [resources]
        self.resources += resources
        return self

    def not_on(self, resources):
        """Add resource(s) on which not to apply the statement.

        :param resources: resource or list of resources
        :type resources: list[str] | str | GetAtt
        :return: the modified statement
        :rtype: Statement
        """
        assert not self.resources
        if isinstance(resources, str) or isinstance(resources, GetAtt):
            resources = [resources]
        self.not_resources += resources
        return self

    def for_service(self, services):
        """Add service that can use the statement.

        :param services: service of list of services
        :type services: list[str] | str | GetAtt
        :return: the modified statement
        :rtype: Statement
        """
        if isinstance(services, str) or isinstance(services, GetAtt):
            services = [services]
        if 'Service' not in self.principals:
            self.principals['Service'] = []
        self.principals['Service'] += services
        return self

    # Declare an alias
    service = for_service


class Allow(Statement):
    """Allow statement."""

    EFFECT = 'Allow'


class Deny(Statement):
    """Deny statement."""

    EFFECT = 'Deny'


class PolicyDocument(object):
    """IAM Policy Document."""

    def __init__(self):
        """Initialize a policy document."""
        self.statements = []

    def append(self, statement):
        """Append a statement.

        :param statement: a IAM Statement
        :type statement: Statement
        :return: the modified policy document
        :rtype: PolicyDocument
        """
        assert isinstance(statement, Statement)
        self.statements.append(statement)
        return self

    def extend(self, statements):
        """Append a list of statements.

        :param statements: IAM Statements
        :type statements: list[Statement]
        :return: the modified policy document
        :rtype: PolicyDocument
        """
        for s in statements:
            assert isinstance(s, Statement)
        self.statements.extend(statements)
        return self

    def __iadd__(self, statements):
        """see extend."""
        return self.extend(statements)

    def __add__(self, other):
        """Add statement(s) or merge two policy documents.

        :param other: statement, list of statements or policy document
        :type other: Statement | list[Statement] | PolicyDocument
        :return: the modified policy document
        :rtype: PolicyDocument
        """
        result = PolicyDocument()
        result.extend(self.statements)
        if isinstance(other, Statement):
            result.append(other)
        elif isinstance(other, PolicyDocument):
            result.extend(other.statements)
        else:
            result.extend(other)
        return result

    @property
    def properties(self):
        """Serialize the object as a simple dict.

        Can be used to transform to CloudFormation Yaml format.

        :rtype: dict
        """
        assert self.statements, 'A policy should have at least one statement'
        return {'Version': '2012-10-17',
                'Statement': [s.properties for s in self.statements]}


INSTANCE_ASSUME_ROLE = Allow().to(
    'sts:AssumeRole').for_service('ec2.amazonaws.com')


class Policy(Resource):
    """A CloudFormation Policy resource."""

    def __init__(self,
                 name,
                 policy_document=None,
                 roles=None,
                 groups=None,
                 users=None):
        """Initialize a policy.

        :param name: logical name on the stack
        :type name: str
        :param policy_document: policy document
        :type policy_document: PolicyDocument
        :param roles: list of roles to apply the policy to
        :type roles: list[str] | None
        :param groups: list of groupss to apply the policy to
        :type groups: list[str] | None
        :param users: list of users to apply the policy to
        :type users: list[str] | None
        """
        super(Policy, self).__init__(name, kind=AWSType.IAM_POLICY)
        self.roles = roles
        self.groups = groups
        self.users = users
        self.policy_document = policy_document

    @property
    def properties(self):
        """Serialize the object as a simple dict.

        Can be used to transform to CloudFormation Yaml format.

        :rtype: dict
        """
        result = {'PolicyName': self.name}
        if self.policy_document is not None:
            result['PolicyDocument'] = self.policy_document.properties

        if self.roles is not None:
            result['Roles'] = self.roles
        if self.groups is not None:
            result['Groups'] = self.groups
        if self.users is not None:
            result['Users'] = self.users
        return result


class InstanceProfile(Resource):
    """IAM Instance profile."""

    def __init__(self, name, role):
        """Initialize an instance profile.

        :param name: logical name in the stack
        :type name: str
        :param role: name of the associated role
        :type role: str
        """
        super(InstanceProfile, self).__init__(
            name,
            kind=AWSType.IAM_INSTANCE_PROFILE)
        self.role = role

    @property
    def properties(self):
        """Serialize the object as a simple dict.

        Can be used to transform to CloudFormation Yaml format.

        :rtype: dict
        """
        return {'Roles': [self.role],
                'Path': '/'}


class Role(Resource):
    """IAM Role."""

    def __init__(self, name, assume_role_policy, path='/'):
        """Initialize IAM Role.

        :param name: role name
        :type name: str
        :param assume_role_policy: policy to define who can assume the role
        :type assume_role_policy: PolicyDocument
        :param path: the path associated with this role (default: /)
        :type path: str
        """
        # Note: we don't set RoleName attribute because of limitations during
        # update with cloudform. In that case RoleName is generated directly by
        # Cloud Formation. This is not related to the name of the resource as
        # part of a stack.
        super(Role, self).__init__(name, kind=AWSType.IAM_ROLE)
        self.path = path
        self.policies = []
        self.assume_role_policy = assume_role_policy

    def add(self, policy):
        """Add a policy.

        :param policy: a policy
        :type policy: Policy
        :return: the object itself
        :rtype: Role
        """
        assert isinstance(policy, Policy)
        self.policies.append(policy)
        return self

    @property
    def properties(self):
        """Serialize the object as a simple dict.

        Can be used to transform to CloudFormation Yaml format.

        :rtype: dict
        """
        return {'AssumeRolePolicyDocument': self.assume_role_policy.properties,
                'Path': self.path,
                'Policies': [p.properties for p in self.policies]}


class InstanceRole(Stack):
    """Instance Role.

    Create both Role and associated profile.
    """

    def __init__(self, name, path='/'):
        """Initialize an instance role.

        :param name: name of the role. The instance profile name will be this
            name + ``InstanceProfile``
        :type name: str
        :param path: path associated with this role
        :type path: str
        """
        super(InstanceRole, self).__init__(name)
        assume_role_policy = PolicyDocument()
        assume_role_policy.append(INSTANCE_ASSUME_ROLE)
        self.add(Role(name, assume_role_policy, path))
        self.add(InstanceProfile(name + 'InstanceProfile',
                                 self.resources[name].ref))

    def add_policy(self, policy):
        """Add a policy.

        :param policy: a policy
        :type policy: Policy
        :return: the object itself
        :rtype: InstanceRole
        """
        self[self.name].add(policy)
        return self

    @property
    def instance_profile(self):
        """Return the name of the instance profile.

        :return: the name of the instance profile
        :rtype: str
        """
        return self.resources[self.name + 'InstanceProfile']
