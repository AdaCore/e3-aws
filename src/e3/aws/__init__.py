from __future__ import annotations
from typing import TYPE_CHECKING
import argparse
import boto3
import botocore.session
import json
import logging
import os
import re
import requests
import requests.auth
import urllib.parse

from botocore.auth import SigV4Auth
from botocore.awsrequest import AWSRequest
from botocore.stub import Stubber
from uuid import uuid4


from e3.error import E3Error
from e3.env import Env
from e3.os.process import Run

logger = logging.getLogger(__name__)

if TYPE_CHECKING:
    from typing import Any, Dict, List, Optional, Callable
    import botocore.client
    import botocore.stub


class AWSSessionRunError(E3Error):
    def __init__(
        self, message: str, origin: str, process: Optional[Run] = None
    ) -> None:
        """Initialize an AWSSessionRunError.

        :param message: the exception message
        :param origin: the name of the function, class, or module having raised
        :param process: process that failed
        """
        super().__init__(message, origin)
        self.origin = origin
        self.message = message
        self.process = process


class Session:
    """Handle AWS session and clients."""

    def __init__(
        self,
        regions: Optional[list[str]] = None,
        stub: bool = False,
        profile: Optional[str] = None,
        credentials: Optional[Dict] = None,
    ) -> None:
        """Initialize an AWS session.

        Once initialized AWS environment can be accessed from Env().aws_env

        :param regions: list of regions to work on. The first region is
            considered as the default region. This parameter should be provided
            if AWS environment variables are not used to specified the region
        :param stub: if True clients are necessarily stubbed
        :param profile: profile name
        :param credentials: AWS credentials dictionary containing the
            following keys: AccessKeyId, SecretAccessKey, SessionToken
            as returned by ``assume_role``
        """
        if profile is not None or credentials is None:
            self.session = botocore.session.Session(profile=profile)
        else:
            self.session = botocore.session.Session()
            self.session.set_credentials(
                access_key=credentials["AccessKeyId"],
                secret_key=credentials["SecretAccessKey"],
                token=credentials["SessionToken"],
            )

        self.profile = profile
        if regions is None:
            # the value return below is ('region', 'AWS_DEFAULT_REGION', None, None)
            # See botocore/configprovider.py
            region_variable = self.session.SESSION_VARIABLES["region"][1]
            region = os.environ.get(region_variable, "")
            if not region:
                raise ValueError(
                    "region should be specified either using regions "
                    "parameter or using AWS environment variables"
                )
            self.regions = [region]
        else:
            self.regions = regions

        self.default_region = self.regions[0]

        self.force_stub = stub
        self.clients: dict[str, botocore.client.Client] = {}
        self.stubbers: dict[str, botocore.stub.Stubber] = {}

        self._account_alias = None

        self._identity = None

    def assume_role(
        self,
        role_arn: str,
        role_session_name: str,
        session_duration: Optional[int] = None,
    ) -> Session:
        """Return a session with ``role_arn`` credentials.

        :param role_arn: ARN of the role to assume
        :param role_session_name: a name to associate with the created
            session
        :param session_duration: session duration in seconds or None for
            default
        :return: a Session instance
        :rtype: Session
        """
        credentials = self.assume_role_get_credentials(
            role_arn, role_session_name, session_duration=session_duration
        )
        return Session(regions=self.regions, credentials=credentials)

    def assume_role_get_credentials(
        self,
        role_arn: str,
        role_session_name: str,
        session_duration: Optional[int] = None,
        as_env_var: bool = False,
    ) -> Dict[str, Any]:
        """Return credentials for ``role_arn``.

        :param role_arn: ARN of the role to assume
        :param role_session_name: a name to associate with the created
            session
        :param session_duration: session duration in seconds or None for
            default
        :param as_env_var: if set to True the returned credentials dictionnary
            keys are translated to be compatible to update os.environ.
        """
        client = self.client("sts", region=self.regions[0])
        arguments: Dict[str, Any] = {
            "RoleArn": role_arn,
            "RoleSessionName": role_session_name,
        }
        if session_duration is not None:
            arguments["DurationSeconds"] = session_duration

        # If AWS_MFA_DEVICE is in the environment ask user for OTP
        if "AWS_MFA_DEVICE" in os.environ:  # all: no cover
            arguments["SerialNumber"] = os.environ["AWS_MFA_DEVICE"]
            otp = input("Enter MFA code: ")
            arguments["TokenCode"] = otp

        response = client.assume_role(**arguments)

        credentials = response["Credentials"]
        if as_env_var:
            key_to_envvar = {
                "AccessKeyId": "AWS_ACCESS_KEY_ID",
                "SecretAccessKey": "AWS_SECRET_ACCESS_KEY",
                "SessionToken": "AWS_SESSION_TOKEN",
            }
            credentials = {
                key_to_envvar[k]: v
                for k, v in credentials.items()
                if k in key_to_envvar
            }

        return credentials

    @property
    def account_alias(self):
        """Return current account alias."""
        if self._account_alias is None:
            client = self.client("iam", region="us-east-1")
            aliases = client.list_account_aliases()["AccountAliases"]
            if aliases:
                # Even if the API return a list there currently only one
                # alias possible
                self._account_alias = aliases[0]
            else:
                self._account_alias = ""
        return self._account_alias

    @property
    def identity(self):
        """Return identity information."""
        if self._identity is None:
            sts_client = self.client("sts")
            self._identity = {"UserId": "", "Account": "", "Arn": ""}
            self._identity.update(sts_client.get_caller_identity())

        return self._identity

    def stub(self, name: str, region: Optional[str] = None) -> botocore.stub.Stubber:
        """Return stub for a given client.

        Note that if the client does not exist yet it will be created.

        :param name: client name
        :param region: region associated with the client. If None the default
            region is taken.
        :return: the stub instance
        """
        if not self.force_stub:
            return None
        if region is None:
            region = self.default_region

        if name not in self.stubbers or region not in self.stubbers[name]:
            # Create client
            self.client(name, region)

        return self.stubbers[name][region]

    def to_boto3(self) -> boto3.Session:
        """Return boto3 session initialized from current botocore session."""
        credentials = self.session.get_credentials()
        frozen_credentials = credentials.get_frozen_credentials()
        return boto3.Session(
            aws_access_key_id=frozen_credentials.access_key,
            aws_secret_access_key=frozen_credentials.secret_key,
            aws_session_token=frozen_credentials.token,
            region_name=self.default_region,
            profile_name=self.profile,
        )

    def client(self, name: str, region: Optional[str] = None) -> botocore.client.Client:
        """Get a client.

        :param name: client name
        :param region: region associated with the client. If None the default
            region is taken.
        :return: a client instance
        """
        if region is None:
            region = self.default_region

        assert region is not None, "no region or default_region set"

        if name not in self.clients:
            self.clients[name] = {}
            self.stubbers[name] = {}

        if region not in self.clients[name]:
            self.clients[name][region] = self.session.create_client(
                name, region_name=region
            )
            if self.force_stub:
                self.stubbers[name][region] = Stubber(self.clients[name][region])
                self.stubbers[name][region].activate()

        return self.clients[name][region]

    def run(
        self, cmd: List[str], role_arn: str, session_duration: int, **kwargs: Any
    ) -> Run:
        """Execute a command with credentials to assume role role_arn.

        :param cmd: command to execute
        :role_arn: Arn of the role to be used by the command
        :session_duration: session duration in seconds or None for default
        :param kwargs: additional parameters to provide to e3.os.process.Run
        :return: Result of the call to Run for the command
        """
        credentials = self.assume_role_get_credentials(
            role_arn, "aws_run_session", session_duration, as_env_var=True
        )

        if "env" not in kwargs:
            if "ignore_environ" not in kwargs:
                kwargs["ignore_environ"] = False
            kwargs["env"] = credentials
        else:
            kwargs["env"] = dict(kwargs["env"]).update(credentials)

        aws_p = Run(cmd, **kwargs)

        if aws_p.status:
            raise AWSSessionRunError(
                f"{cmd} failed (exit status: {aws_p.status})",
                origin="aws_session_cli_cmd",
                process=aws_p,
            )
        return aws_p


class AWSEnv(Session):
    """Handle AWS session and clients."""

    def __init__(
        self,
        regions: Optional[list[str]] = None,
        stub: bool = False,
        profile: Optional[str] = None,
    ):
        """Initialize an AWS session.

        Once intialized AWS environment can be accessed from Env().aws_env

        :param regions: list of regions to work on. The first region is
            considered as the default region.
        :param stub: if True clients are necessarily stubbed
        :param profile: profile name
        """
        super().__init__(regions=regions, stub=stub, profile=profile)
        env = Env()
        env.aws_env = self


class default_region:
    """Context manager used to set a default region."""

    def __init__(self, region: str):
        """Initialize context manager.

        :param region: default region
        """
        aws_env = Env().aws_env

        self.previous_region = aws_env.default_region
        self.default_region = region

    def __enter__(self):
        Env().aws_env.default_region = self.default_region

    def __exit__(self, _type, _value, _tb):
        del _type, _value, _tb
        Env().aws_env.default_region = self.previous_region


def session() -> Callable:
    """Decorate a function to handle automatically AWS session retrieval.

    The function in input should take a mandatory argument called session.

    :param name: session name
    :type name: str
    """

    def decorator(func):
        def wrapper(*args, **kwargs):
            if "session" in kwargs:
                session = kwargs.get("session")
                del kwargs["session"]
            else:
                session = Env().aws_env
            return func(*args, session=session, **kwargs)

        return wrapper

    return decorator


def assume_role_main():
    """Generate shell commands to set credentials for a role."""
    argument_parser = argparse.ArgumentParser()
    argument_parser.add_argument(
        "--json", action="store_true", help="Output credentials as JSON"
    )
    argument_parser.add_argument(
        "--role-session-name",
        help="Role session name, by default a random value is generated",
    )
    argument_parser.add_argument("--session-duration")
    argument_parser.add_argument("role_arn")
    args = argument_parser.parse_args()

    key_to_envvar = {
        "AccessKeyId": "AWS_ACCESS_KEY_ID",
        "SecretAccessKey": "AWS_SECRET_ACCESS_KEY",
        "SessionToken": "AWS_SESSION_TOKEN",
    }

    s = Session(regions=["eu-west-1"])
    session_duration = args.session_duration
    if session_duration is not None:
        session_duration = int(session_duration)

    if args.role_session_name:
        role_session_name = args.role_session_name
    else:
        role_session_name = str(uuid4()).replace("-", "")

    credentials = s.assume_role_get_credentials(
        args.role_arn, role_session_name, session_duration=session_duration
    )
    credentials["Expiration"] = credentials["Expiration"].timestamp()
    if args.json:
        print(json.dumps(credentials))
    else:
        credentials = {
            key_to_envvar[k]: v for k, v in credentials.items() if k in key_to_envvar
        }
        for k, v in credentials.items():
            print(f"export {k}={v}")


def iterate(fun: Callable, key: str, **kwargs: Any) -> Any:
    """Create an iterator other paginate botocore function.

    :param fun: the function to call
    :param key: the key in the returned data containing the elements
    :param kwargs: parameters passed to the function
    """
    result = fun(**kwargs)
    for data in result.get(key, []):
        yield data

    while result.get("NextToken"):
        result = fun(NextToken=result["NextToken"], **kwargs)
        for data in result.get(key, []):
            yield data


def name_to_id(name: str) -> str:
    """Convert a resource name to a resource id.

    The conversion is done by removing characters that are not alphanumeric as
    resource id must contain only alphanumeric character and first character and
    characters following a dash to uppercase for a better readability.
    """

    def replacement(match):
        """Return uppercased second character of the match."""
        return match.group(1)[1].upper()

    resource_id = re.sub(r"[^a-zA-Z0-9]", "", re.sub(r"(-[a-z])", replacement, name))
    resource_id = resource_id[0].upper() + resource_id[1:]
    return resource_id


class IAMAuth(requests.auth.AuthBase):
    """Authorizer for the requests framework that use AWS Signature V4 protocol."""

    def __init__(
        self, session: Session, role: Optional[str] = None, region: Optional[str] = None
    ):
        """Initialize authorizer.

        :param session: an aws session
        :param role: if not None role to assume
        :param region: if None use default region (from session) otherwise use it as
            region
        """
        self.session = session
        self.role = role
        if region is None:
            self.region = self.session.default_region
        else:
            self.region = region

    def __call__(self, request: requests.PreparedRequest) -> requests.PreparedRequest:
        """See requests framework."""
        # The code does not implement the signature v4 protocol. It reuses parts of
        # botocore to do so. Botocore does not provide a generic interface to sign
        # a given request. So the following code create a temporary AWSRequest object
        # with the same parameters as the user request, sign it and then extract
        # the signature (i.e: the headers) that is re-injected into the user request.
        session = self.session
        if self.role is not None:
            session = self.session.assume_role(self.role, "iamauthsession")

        credentials = session.session.get_credentials().get_frozen_credentials()

        # Split back the url in order to be able to call AWSRequest
        aws_headers = dict(request.headers)
        key_to_delete = []
        for key in aws_headers:
            if key.lower() in ("accept", "accept-encoding", "connection"):
                key_to_delete.append(key)

        for key in key_to_delete:
            del aws_headers[key]

        parsed_url = urllib.parse.urlparse(request.url)
        aws_params = {
            str(k): ",".join(v)
            for k, v in urllib.parse.parse_qs(str(parsed_url.query)).items()
        }
        aws_url = urllib.parse.urlunparse(
            (
                str(parsed_url.scheme),
                str(parsed_url.netloc),
                str(parsed_url.path),
                str(parsed_url.params),
                "",
                "",
            )
        )

        # Compute the headers for the request
        aws_request = AWSRequest(
            method=request.method, url=aws_url, params=aws_params, headers=aws_headers
        )
        SigV4Auth(credentials, "execute-api", self.region).add_auth(aws_request)

        # Update request headers
        request.headers.update(aws_request.headers)
        return request
