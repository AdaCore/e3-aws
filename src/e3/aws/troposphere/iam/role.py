"""Provide IAM Role classes."""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from troposphere import AWSObject, iam, Tags, GetAtt

from e3.aws import name_to_id
from e3.aws.troposphere import Construct
from e3.aws.troposphere.iam.policy_document import PolicyDocument
from e3.aws.troposphere.iam.policy_statement import AssumeRole, Trust, Allow

if TYPE_CHECKING:
    from typing import Optional
    from e3.aws.troposphere import Stack
    from e3.aws.troposphere.iam.policy_statement import PrincipalType


@dataclass
class Role(Construct):
    """Define IAM role with an attached assume_role policy document.

    :param name: role name
    :param description: role description
    :param trust: trust policy. It can be either a principal, a Trust statement or
        a policy document
    :param managed_policy_arns: list of ARNs of IAM managed policies to attach
        to the role
    :param max_session_duration: the maximum session duration (in seconds) that
        you want to set for the specified role. default is one hour
    :param tags: a list of tags that are attached to the specified role
    """

    name: str
    description: str
    trust: PrincipalType | Trust | PolicyDocument
    managed_policy_arns: Optional[list[str]] = None
    max_session_duration: Optional[int] = None
    tags: dict[str, str] = field(default_factory=lambda: {})
    path: str = "/"
    boundary: Optional[str] = None

    @property
    def trust_policy(self) -> PolicyDocument:
        """Return the trust policy for the role."""
        if isinstance(self.trust, Trust):
            return PolicyDocument(statements=[self.trust])
        elif isinstance(self.trust, PolicyDocument):
            return self.trust
        else:
            return PolicyDocument(statements=[AssumeRole(principal=self.trust)])

    @property
    def arn(self):
        return GetAtt(name_to_id(self.name), "Arn")

    def resources(self, stack: Stack) -> list[AWSObject]:
        """Return troposphere objects defining the role."""
        attr = {}

        for key, val in {
            "RoleName": self.name,
            "Description": self.description,
            "ManagedPolicyArns": self.managed_policy_arns,
            "MaxSessionDuration": self.max_session_duration,
            "AssumeRolePolicyDocument": self.trust_policy.as_dict,
            "Tags": Tags({"Name": self.name, **self.tags}),
            "Path": self.path,
            "PermissionsBoundary": self.boundary,
        }.items():
            if val is not None:
                attr[key] = val

        return [iam.Role(name_to_id(self.name), **attr)]

    def cfn_policy_document(self, stack: Stack) -> PolicyDocument:
        return PolicyDocument(
            statements=[
                Allow(
                    action=[
                        "iam:GetRole",
                        "iam:CreateRole",
                        "iam:AttachRolePolicy",
                        "iam:DetachRolePolicy",
                        "iam:DeleteRole",
                    ],
                    resource=f"arn:aws:iam::%(account)s:role/{self.name}",
                )
            ]
        )
