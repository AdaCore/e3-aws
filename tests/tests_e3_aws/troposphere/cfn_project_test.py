"""Provide Cloudformation construct tests."""

from __future__ import annotations

import os
from typing import TYPE_CHECKING

from e3.aws import AWSEnv
from e3.aws.troposphere import CFNProjectMain
from e3.aws.troposphere.iam.role import Role


if TYPE_CHECKING:
    from e3.aws.troposphere import Stack


TEST_DIR = os.path.dirname(os.path.abspath(__file__))


class MyCFNProject(CFNProjectMain):
    """Provide CLI to manage MyCFNProject."""

    def create_stack(self) -> list[Stack]:
        """Return MyCFNProject stack."""
        self.add(
            (
                Role(
                    name="TestRole",
                    description="TestRole description",
                    trust={"Service": "test"},
                )
            )
        )
        return self.stack


def test_cfn_project_main(capfd) -> None:
    """Test CFNProjectMain."""
    aws_env = AWSEnv(regions=["eu-west-1"], stub=True)
    test = MyCFNProject(
        name="TestProject",
        account_id="12345678",
        stack_description="TestStack",
        s3_bucket="cfn-test-deploy-bucket",
        regions=["eu-west-1"],
    )
    test.execute(args=["show"], aws_env=aws_env)

    captured = capfd.readouterr()
    print(captured.out)
    with open(os.path.join(TEST_DIR, "cfn_project_test.out")) as f_out:
        assert captured.out == f_out.read()
