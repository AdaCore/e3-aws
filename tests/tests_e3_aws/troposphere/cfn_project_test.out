AWSTemplateFormatVersion: '2010-09-09'
Description: TestStack
Resources:
  TestRole:
    Properties:
      AssumeRolePolicyDocument:
        Statement:
        - Action: sts:AssumeRole
          Effect: Allow
          Principal:
            Service: test
        Version: '2012-10-17'
      Description: TestRole description
      Path: /
      RoleName: TestRole
      Tags:
      - Key: Name
        Value: TestRole
    Type: AWS::IAM::Role

